kimchi/src/circuits/argument.rs

```diff
@@ -33,13 +39,24 @@ pub trait Argument<F: FftField> {
     /// The number of constraints created by the argument.
     const CONSTRAINTS: u32;
 
+    /// Constraints for this argument
+    fn constraints<T: ArithmeticOps<F>>(witness: &GateWitness<T>, constants: Vec<T>) -> Vec<T>;
```

Comment:
btw have we considered implementing `Field` on the types in `expr.rs` instead of creating an `ArithmeticOps`? Could be cleaner

Reply:
also semantically I don't understand this change, why would a gate take a witness?

Reply:
It's an idea I have considered, but Field seems to have a lot of other things we would not want to implement.  I will consider again and post back here.

Reply:
> why would a gate take a witness?
The gate operates on the witness to see whether it's relation is satisfied by the witness assignment.  This allows the definition of the gate, i.e., its constraints to be reused for witness verification.  That is we can have a generic `verify_witness()` that can not only reuse the constraints (no second implementation required), but also even pinpoint which constraints failed, thus, aiding the debug process for gate developers.

Reply:
> also semantically I don't understand this change, why would a gate take a witness?


The gate operates on the witness to see whether its relation is satisfied by the witness assignment.  This allows the definition of the gate (i.e., its constraints) to be reused for witness verification (not only expr generation).  That is, we can have a generic `verify_witness()` that can not only reuse the constraints (no second implementation required), but also even pinpoint which constraints failed, thus, aiding the debug process for gate developers.

Reply:
> It's an idea I have considered, but Field seems to have a lot of other things we would not want to implement. I will consider again and post back here.

yeah but we can easily just add `unimplemented!()` for all the functions we don't care about in the implementation. It could make a lot of this code cleaner potentially

Reply:
no - it's a work in progress, but I put it here to get early feedback.

Hmm interesting idea to use `unimplemented!()`, but not only would we need to implement `Field` on `expr.rs`, but we need some expression functions' behaviour to change when used on real field elements instead of expressions.

Reply:
n.b. we are not mixing test code with production.  We are making it so that we can directly test the some of the production code without running the prover or needing to create an alternative implementation to check the witness.